polkadot/xcm/xcm-builder/src/routing.rs

```diff
@@ -139,3 +139,69 @@ impl EnsureDelivery for Tuple {
 		(None, None)
 	}
 }
+
+/// A wrapper router that attempts to *encode* and *decode* passed XCM `message` to ensure that the
+/// receiving side will be able to decode, at least with the same XCM version.
+///
+/// This is designed to be at the top-level of any routers which do the real delivery. While other
+/// routers can manipulate the `message`, we cannot access the final XCM due to the generic
+/// `Inner::Ticket`. Therefore, this router aims to validate at least the passed `message`.
+///
+/// NOTE: For use in mock runtimes which don't have the DMP/UMP/HRMP XCM validations.
+pub struct EnsureDecodableXcm<Inner>(sp_std::marker::PhantomData<Inner>);
+impl<Inner: SendXcm> SendXcm for EnsureDecodableXcm<Inner> {
+	type Ticket = Inner::Ticket;
+
+	fn validate(
+		destination: &mut Option<Location>,
+		message: &mut Option<Xcm<()>>,
+	) -> SendResult<Self::Ticket> {
+		if let Some(msg) = message {
+			if let Some(e) = Self::ensure_encode_decode(msg) {
```

Comment:
Isn't this already covered by `validate_xcm_nesting()` in https://github.com/paritytech/polkadot-sdk/pull/4236?

It already does encode+decode (with limits) inside HRMP, DMP, UMP.

I believe ^ covers everything, no? Do we still need `EnsureDecodableXcm` wrapper for anything extra?

Reply:
when https://github.com/paritytech/polkadot-sdk/pull/4236 is merged, I plan to remove `fn ensure_encode_decode` and replace it with `VersionedXcm::<()>::from(msg.clone()).validate_xcm_nesting()` here.

yes, #4236 covers real cases for HRMP, DMP, UMP with real runtimes, but still does not cover cases for lots of testing/example/mocking stuff, where we use just "dummy" routers

Reply:
Important benchmarks fixes were moved to the #4236, so this PR contains now just `EnsureDecodableXcm` feature for XCM mocks/tests/samples where we don't use HRMP,UMP,DMP routers.

@acatangiu merge or close :D? I am fine with both :)